### ISSUE TITLE
Single + Multiple Download for Images

### ISSUE THREAD

**Issue by asuresh-code** [ral-facilities/object-storage-api#81]
Mirroring this [front-end issue](https://github.com/ral-facilities/inventory-management-system/issues/1078), the back-end needs to allow for images to be downloaded instead of displayed just inline.

Also in line with this feature, file names need to be parsed against their file type to ensure the correct file extension is in place; otherwise there are issues trying to open downloaded images. There are 3 main solutions:

1. **Store Image MIME type in separate column**
2. **Store Image file extension in separate column**
3.  **Store image file extension with filename**

In this PR I've opted for the 1st option. Options 1 and 2 (Storing an extra column) would also allow us to show the image type in the image info, and allow filtering by type. They also seem more robust than Option 3, which would involve string manipulation to remove and add the file extension back in after image edits. 

Between Option 1 and 2, option 2 would be better if we did not want additional enhancements (like viewing and filtering by image type); With Option 2, you would extract the MIME type from the header, convert it to an extension once and store it, where it would be appended to the download url's filename. Whereas Option 1, you would store the MIME type and convert it to an extension everytime an image was downloaded. However, if the filtering/viewing enhancements were implemented it would make more sense to use Option 1.

All of the options are viable however, so any of the implementations work fine.

**Comment by joelvdavies** [comment]
I agree option 1 is more sensible out of these. Just not quite sure why this is needed? The content type is already stored in the object store so I was under the impression all that was needed was an additional parameter in the presigned url to make it download. The content type actually comes from the extension anyway, as the front end determines that when it uploads.

**Comment by asuresh-code** [comment]
> The content type is already stored in the object store so I was under the impression the most that was needed was an additional parameter in the presigned url to make it download

Although the content-type is set in the object store, it doesn't actually seem to affect the file type of the image downloaded (at least on a Windows System). The file extension seems to override what the perceived file type is, so if you download a file called `image.mp4` and set the content type to `"image/jpeg"` both on upload and download, it will still be perceived as an MP4 file. I don't think it's an issue with our implementation, it may be a quirk of generating a presigned url instead of just getting the object.

> download but I even thought that may be possible on the front end as the data is returned from the view anyway

I think you're right for this case; I believe you can set the `ResponseContentDisposition:` as `inline` for both viewing and downloading images, but I didn't investigate that yet. However, setting it to `attachment` for both viewing and downloading images, seems to let you do both as well.

> The content type actually comes from the extension anyway, as the front end determines that when it uploads.

Yes which was actually part of the problem initially. On the Uppy upload dialog, users could edit the file extension (which was part of the filename), renaming a file called image.jpeg to image.mp4, creating the whole problem. I've changed it so now on the uppy dialog the extension is hidden when editing the file name, so that it's not tampered with. If you leave the file extension with the file name, it is not a clean solution since users can edit the filename, so it requires hiding the extension again and removing/adding it back on the back-end (I think Viktor pointed out that it should have verification/validation on the back end in case you are just using the endpoints w/o the API). At that point it's as complicated as Option's 1 and 2, and the former seem more sensible.

**Comment by joelvdavies** [comment]
> Although the content-type is set in the object store, it doesn't actually seem to affect the file type of the image downloaded (at least on a Windows System). The file extension seems to override what the perceived file type is, so if you download a file called `image.mp4` and set the content type to `"image/jpeg"` both on upload and download, it will still be perceived as an MP4 file. I don't think it's an issue with our implementation, it may be a quirk of generating a presigned url instead of just getting the object.
> 
Yes this is just how file extensions behave, I think the original behaviour is expected and I personally don't think it needs preventing here. You can always rename a .png to a .pdf and try and upload it and the uploader would assume its a pdf event thought the data itself is not.
> 
> I think you're right for this case; I believe you can set the `ResponseContentDisposition:` as `inline` for both viewing and downloading images, but I didn't investigate that yet. However, setting it to `attachment` for both viewing and downloading images, seems to let you do both as well.
> 
Useful to know, I thought we would want inline because that is what we use when viewing, but I would also want to avoid having a separate endpoint to download by handling it on the front end.
> 
> Yes which was actually part of the problem initially. On the Uppy upload dialog, users could edit the file extension (which was part of the filename), renaming a file called image.jpeg to image.mp4, creating the whole problem. I've changed it so now on the uppy dialog the extension is hidden when editing the file name, so that it's not tampered with. 

Yeah I like that solution.

> If you leave the file extension with the file name, it is not a clean solution since users can edit the filename, so it requires hiding the extension again and removing/adding it back on the back-end (I think Viktor pointed out that it should have verification/validation on the back end in case you are just using the endpoints w/o the API).  At that point it's as complicated as Option's 1 and 2, and the former seem more sensible.

I agree that is the normal approach, but given that its not any different to renaming a file name's extension, I'm not sure its worth duplicating the content type in the database.

**Comment by asuresh-code** [comment]
> Yes this is just how file extensions behave, I think the original behaviour is expected and I personally don't think it needs preventing here. You can always rename a .png to a .pdf and try and upload it and the uploader would assume its a pdf event thought the data itself is not.

Yeah, this was one of the first questions we asked; whether it needs preventing in the first place. It was a close decision to address it, but I think one of the main reasons why we decided to is because users were able to edit the file extension (which I believe you can't in most file systems), and the solution didn't seem to difficult to implement.

> I agree that is the normal approach, but given that its not any different to renaming a file name's extension, I'm not sure its worth duplicating the content type in the database.

Yeah, I think the only "functional" reason to keep the content type/file extension separate is if you wanted to display/filter by content type. You could probably derive the file type from the file name w/ the extension, it would just be a bit more complicated. 

> Maybe its worth discussing further and addressing in a separate task.

Yes further discussion would probably be beneficial.
